KerbalStuff/common.py

```diff
@@ -150,18 +152,95 @@ def get_page() -> int:
         return 1
 
 
-def get_mods(ga: Game = None, query: str = '', page_size: int = 30) -> Tuple[Iterable[Mod], int, int]:
+def get_paginated_mods(ga: Game = None, query: str = '', page_size: int = 30) -> Tuple[Iterable[Mod], int, int]:
     page = get_page()
-    mods, total_pages = search_mods(ga, query, page, page_size)
+    mods, total_pages = search_mods(ga.id if ga else None, query, page, page_size)
     return mods, page, total_pages
 
 
+# (2 games + None) * (limit 6 + limit 30) = 6 argument combinations
+@cached(cache=TTLCache(maxsize=10, ttl=1800))
+def get_featured_mods(game_id: Optional[int], limit: int) -> List[Mod]:
+    mods = Featured.query.outerjoin(Mod).filter(Mod.published).order_by(desc(Featured.created))
+    if game_id:
+        mods = mods.filter(Mod.game_id == game_id)
+    return mods.limit(limit).all()
+
+
+@cached(cache=TTLCache(maxsize=10, ttl=1800))
+def get_top_mods(game_id: Optional[int], limit: int) -> List[Mod]:
+    mods = Mod.query.filter(Mod.published).order_by(desc(Mod.score))
+    if game_id:
+        mods = mods.filter(Mod.game_id == game_id)
+    return mods.limit(limit).all()
+
+
+@cached(cache=TTLCache(maxsize=10, ttl=1800))
+def get_new_mods(game_id: Optional[int], limit: int) -> List[Mod]:
+    mods = Mod.query.filter(Mod.published).order_by(desc(Mod.created))
+    if game_id:
+        mods = mods.filter(Mod.game_id == game_id)
```

Comment:
Do you know whether sqlalchemy or postgresql is smart enough to apply this filter before we sort the rows?

Reply:
This should be handled by the Postgres query planner. In theory it doesn't even need to do much sorting, since we have an index on `Mod.created`  which it can just walk through.
I'm gonna check the query sent to Postgres though, and see if reordering the statements changes something.

Reply:
As I suspected, the query that SqlAlchemy sends to Postgres is always formatted the same, regardless of code order:
```
SELECT mod.id AS mod_id, mod.created AS mod_created, mod.updated AS mod_updated, mod.user_id AS mod_user_id, mod.game_id AS mod_game_id, mod.name AS mod_name, mod.description AS mod_description, mod.short_description AS mod_short_description, mod.published AS mod_published, mod.locked AS mod_locked, mod.locked_by_id AS mod_locked_by_id, mod.lock_reason AS mod_lock_reason, mod.donation_link AS mod_donation_link, mod.external_link AS mod_external_link, mod.license AS mod_license, mod.votes AS mod_votes, mod.score AS mod_score, mod.background AS mod_background, mod."bgOffsetX" AS "mod_bgOffsetX", mod."bgOffsetY" AS "mod_bgOffsetY", mod.default_version_id AS mod_default_version_id, mod.source_link AS mod_source_link, mod.follower_count AS mod_follower_count, mod.download_count AS mod_download_count, mod.ckan AS mod_ckan
FROM mod
WHERE mod.published AND mod.game_id = 2
ORDER BY mod.created DESC
LIMIT 30
```
and an EXPLAIN ANALYZE on the query shows that it just steps through the index backwards, filtering out rows with the wrong game id or `published = false`, and stopping after `limit` rows:
![image](https://user-images.githubusercontent.com/28812678/117568266-ab0e1100-b0bf-11eb-8902-fccfefde73df.png)

Reply:
Excellent, that's great news, thanks for looking into it!